Implement faster ThreadContextMap

jengebr opened the pull request:
This implements a version of `ThreadContextMap` that is faster than `DefaultThreadContextMap, and is suitable for applications with nested classloaders, such as web applications running in containers.  Key characteristics of this version are:

1. Faster copy-and-modify operations
2. Lower memory allocation
3. State is contained in a single object using a JVM-defined class (`Object[]`).  This solves the classloader reference problem.
4. Performance is better than `DefaultThreadContextMap` except for `get()` operations, where performance becomes linear.

See PR #2321 for additional context and related conversations.

jengebr commented:
@jvz not sure what your schedule is, but I'll be unavailable next week and unable to respond to comments.  I've lined up a coworker to watch the conversation and help where possible.

jengebr commented:
Our application ran into a [JDK performance issue](https://bugs.openjdk.org/browse/JDK-8324186) which makes constructors slow whenever the object has a final member variable.  We're removing finals from hotspots, and the code in this PR contains several instances of this problem.  Any objection if we remove the keyword final from those member variables?  Not sure if it'll violate enforced styling.

vy commented:
@jengebr, very sharp of you! Thanks for chasing this! Keeping `final`s out due to the reason you stated is perfectly fine, but please document it (preferably, with a pointer to the relevant OpenJDK issue) wherever this modification is needed.

OTOH, given the impact surface of this problem and its severity, I expect it to be promptly addressed by OpenJDK. Maybe we should put a reminder to remove this in a year or so. WDYT?

ppkarwasz reviewed (changes requested):
@jengebr,

This seems as a very promising successor of `DefaultThreadContextMap`. I would just:

- repackage it to `org.apache.logging.log4j.internal.map` (or something similar). This package will **not** be exported from the JPMS module, so it can not be used outside of `log4j-api`. If people show interest in using this implementation outside of Log4j API, you can resubmit it to Commons Collections,
- remove the support for inheritable `ThreadLocal`s. IMHO it is an anachronistic way to propagate contexts between threads, that might have been useful, when people used `new Thread` directly.

Regarding the second point I thought about introducing soon (after the `2.23.1` release) two methods to `ThreadContextMap`:

```java
    /**
     * Saves the current context map data.
     * <p>
     *     The returned object is not affected by later changes to the context map.
     * </p>
     * @return An opaque immutable version of the internal context map representation used.
     * @see #restoreContextMap
     * @since 2.24
     */
    Object saveContextMap();

    /**
     * Restores the context map data, from a saved version.
     * @param contextMap An opaque immutable version of the internal context map representation used.
     * @see #saveContextMap
     * @since 2.24
     */
    void restoreContextMap(Object contextMap);
```

These methods will solve two problems:

1. Will allow us to enter the modern era of context propagation. For example we can use them to easily implement `ThreadLocalAccessor` from the [Micrometer Context Propagation](https://docs.micrometer.io/context-propagation/reference/purpose.html) library,
2. Will solve your original problem with `CloseableThreadContext`: we can use `saveContextMap/restoreContextMap` to efficiently restore the initial context data.

jengebr commented on log4j-api/src/main/java/org/apache/logging/log4j/spi/StringArrayThreadContextMap.java:
```diff
@@ -0,0 +1,253 @@
+     */
+    public static final String INHERITABLE_MAP = "isThreadContextMapInheritable";
+
+    private final boolean useMap;
+    private final ThreadLocal<Object[]> threadLocalMapState;
+
+    private static boolean inheritableMap;
```
Thanks for the feedback!  Questions:
1. Can you please clarify inheritableMap?  It's only used for inheritable thread locals that I will remove.
2. Do you mind if we do the new methods as a separate PR?  It's a good idea but this is already complex.

ppkarwasz replied:
I think you can safely remove both the `useMap` and `inheritableMap` from this PR, they should slim down the code from useless overhead.

jengebr commented on log4j-api/src/main/java/org/apache/logging/log4j/spi/StringArrayThreadContextMap.java:
```diff
@@ -0,0 +1,253 @@
+    // (This method is package protected for JUnit tests.)
+    static ThreadLocal<Object[]> createThreadLocalMap(final boolean isMapEnabled) {
+        if (inheritableMap) {
+            return new InheritableThreadLocal<Object[]>() {
+                @Override
+                protected Object[] childValue(final Object[] parentValue) {
+                    return parentValue;
+                }
+            };
```
Good call on the InheritableThreadLocal, I think you're right that would be another leak.

jengebr commented on log4j-api/src/main/java/org/apache/logging/log4j/spi/UnmodifiableArrayBackedMap.java:
```diff
@@ -0,0 +1,486 @@
+    public boolean containsKey(Object key) {
+        int hashCode = key.hashCode();
+        for (int i = 0; i < numEntries; i++) {
+            if (backingArray[getArrayIndexForKey(i)].hashCode() == hashCode
+                    && backingArray[getArrayIndexForKey(i)].equals(key)) {
```
My JMH test shows that hashcode is faster, but that's because the particular strings are reused, and each String instance caches its hashcode.  In a situation with un-cached hashcodes, you're right that equals() is the better choice.

Our applications use fixed/reusable/cached strings but they might not be the general case.  Which direction should we bias?

equals() is still fine on our apps, no worries if that's your preference.

ppkarwasz replied:
Good point, I think most strings will be compile-time constants, so let's keep `hashCode`.

jengebr commented:
> @jengebr, very sharp of you! Thanks for chasing this! Keeping `final`s out due to the reason you stated is perfectly fine, but please document it (preferably, with a pointer to the relevant OpenJDK issue) wherever this modification is needed.
> 
> OTOH, given the impact surface of this problem and its severity, I expect it to be promptly addressed by OpenJDK. Maybe we should put a reminder to remove this in a year or so. WDYT?

@vy I think the reminder is a very good idea.  How do you suggest we do that?  Is there a github mechanism, or just a comment in the code?

jengebr commented:
I've pushed the updated code, working through unrelated build failures.  May revise further if something actually is related.

jengebr commented:
Local build completed successfully with no further revisions.  I'll be "back" in about 1.5 weeks, will address comments quickly after that.

ppkarwasz commented:
@jengebr,

Thank your for your contribution. I'll enable it by default in the 2.24 branch.

jengebr commented:
Thank you for your encouragement!  I hope this helps other as much as it'll help us.

vy commented:
@jengebr, would you mind reminding me why did we settle on an copy-on-write map instead of a mutable one (and cloning it if `getImmutableMapOrNull()`, `getMap()`, etc. is requested), please?

jengebr commented:
Hi @vy - copy-on-write was the previous state, where it would fully copy the existing HashMap into a new one, then pass that around.  To answer the immediate question, I implemented copy-on-write because that was the current behavior.

As for why it was first introduced, I speculate it's because the Map object itself can be exposed to the caller, and therefore is conceivably at risk of concurrent updates.

  The cost of copying a HashMap is quite high (slow iteration across the original, object allocations to create the new) and the perf advantage of the new one comes from drastically reducing it.  If we dropped the copy-and-modify requirement, we'd see better performance than either HashMap or this.

jengebr commented:
> cloning it if getImmutableMapOrNull(), getMap(), etc. is requested
Currently, direct updates to getMap() are visible to the ThreadContext(), and updates to ThreadContext() are visible to the get*Map() methods.  This would be a breaking change, although possibly an acceptable cost.

vy commented:
@jengebr, thanks so much for the quick and elaborate response! :heart_eyes:

> Currently, direct updates to getMap() are visible to the ThreadContext(), and updates to ThreadContext() are visible to the get*Map() methods.

I guess, while we can still retain these behaviours using a mutable map, I am more inclined to treat these as misuses instead of features.

Right now I am firefighting with #2946 (the issue that we couldn't completely resolve in `2.23.1`) in #3050 – your review is more than welcome! :star_struck: There I will try to just fix the reported bug. I intend to submit another PR rewriting `DefaultThreadContextMap` using a mutable array.

jengebr commented:
Thanks for working on this!  I added one suggestion on the PR (streams add overhead), but I like how much easier it is to read that code.

I look forward to the Mutable array!  Please let me know once it's ready and I can support with some performance analysis.

ppkarwasz commented:
> > Currently, direct updates to getMap() are visible to the ThreadContext(), and updates to ThreadContext() are visible to the get*Map() methods.
> 
> I guess, while we can still retain these behaviours using a mutable map, I am more inclined to treat these as misuses rather than features.

By looking closely at [`ThreadContext.getImmutableContext()`](https://logging.apache.org/log4j/2.x/javadoc/log4j-api/org/apache/logging/log4j/ThreadContext.html#getImmutableContext()) I also believe that the returned object could be mutable and if a user tries to modify it, he breaches the method's contract.

Note that `ThreadContext.getImmutableContext()` is a "feature" of the Log4j API only. SLF4J users can only get a copy using [`MDC.getCopyOfContextMap()`](https://www.slf4j.org/api/org/slf4j/MDC.html#getCopyOfContextMap()), so we don't risk misuse from their part.
